Sequence number based replica allocation

dnhatn opened the pull request:
This change prefers allocating replicas on nodes where it can perform an operation-based recovery or has sync_id match to reduce recovery time. We no longer need to perform a synced_flush in a rolling upgrade or full cluster start with this improvement.

I started with an implementation where I used the persisted global checkpoint from replicas and peer recovery retention leases from primaries to make decisions. However, I was not happy with the extension capturing the persisted global checkpoint (see https://github.com/elastic/elasticsearch/compare/master...dnhatn:replica-allocator-with-gcp?expand=1#diff-275151cc4a5cdf942f310a219e86a403R485).

We don't need the global checkpoint to make decisions for open indices. Having a peer recovery retention lease alone is enough to guarantee to have an operation-based recovery since we [share](https://github.com/elastic/elasticsearch/blob/38277fdca9e7337daeccb1e11fdd107becb780c0/server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java#L936) the persisted global checkpoint between copies. I decided to implement this without the global checkpoint.

I prefer to support closed/frozen indices in a follow-up after we agree on the approach (i.e., using the global checkpoint or the last commit).

Closes #46318

dnhatn committed:
Sequence number based replica allocation

elasticmachine commented:
Pinging @elastic/es-distributed

henningandersen commented on server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java:
```diff
@@ -0,0 +1,137 @@
+        internalCluster().startDataOnlyNode();
+        recoveryStarted.await();
+        nodeWithReplica = internalCluster().startDataOnlyNode(nodeWithReplicaSettings);
+        // AllocationService only calls GatewayAllocator if there're unassigned shards
```
I wonder if this is not important to fix as part of this PR?

If we have an ongoing recovery and we started some node that has potentially better data, we will not use it unless something external triggers it (like here). This could happen for all shards on a node? This was the same so far with the sync id, but I think it is worse now since it makes it less certain that we do ops-based recovery.

And vice-versa, there is a staleness concern too. Consider that a file based recovery has been running for a while and some node containing data has been started for a while too. If someone then triggers this by for instance creating an index, we might cancel the recovery to put it on to the node containing the data. But I think this decision will be made based on stale lease data, since we only clear the async-fetch-cache when shards are started.

dnhatn replied:
> And vice-versa, there is a staleness concern too.

@henningandersen, I opened #47035 using your work in #42518.

> Note that a fix for this might arrive via #46520

Thanks @DaveCTurner.

henningandersen commented on server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java:
```diff
@@ -100,18 +98,24 @@ public void processExistingRecoveries(RoutingAllocation allocation) {
-                    } else {
-                        currentSyncId = null;
-                    }
+                    BooleanSupplier currentNodeCanSkipPhase1 = () -> {
```
I think it would be nicer to just turn this into a regular method?

henningandersen reviewed (commented):
I think this direction looks good. I have left a few initial comments inline, but my main concern is staleness:

I think that the info we have from primary about leases can be stale. As soon as a node with a replica dies, we will reach out to all nodes including primary and read the info. And then cache it until a shard with same shard-id is started. Given the `index.recovery.file_based_threshold`, the staleness may become important in much less time than the default 12h lease expiration.

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java:
```diff
@@ -380,30 +389,63 @@ private static long computeMatchingBytes(TransportNodesListShardStoreMetaData.St
+        // have the persisted global checkpoint from the replica to determine if the last commit is safe. However, the likelihood that
+        // the last commit unsafe is very small. It's probably okay to use the retention lease if the sequence numbers of the last commit
+        // from the primary and replica equal.
+        if (indexMetaData.getState() == IndexMetaData.State.CLOSE ||
```
I don't see that the tests exercise this, but it'd be good if they did.

That said, for closed and frozen indices I think it's ok not to treat them specially and simply to use the retention lease if a retention lease exists. It's going to be very rare that any recovery is needed and if it is we don't really have any option but to perform a file-based recovery anyway, so we may as well do that onto the same node as before.

One issue we might face is that closed indices don't necessarily have a PRRL for every shard copy, and I think this could result in bad behaviour on closed indices where the files don't match (and there's no sync ID). We may need to defer special treatment of closed indices to a followup since I think this'll need a change to `RecoverySourceHandler` to address.

DaveCTurner reviewed (commented):
Great stuff @dnhatn, thanks. I left some points for discussion.

dnhatn commented on server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java:
```diff
@@ -100,18 +98,24 @@ public void processExistingRecoveries(RoutingAllocation allocation) {
+                        }
+                        IndexMetaData indexMetadata = allocation.metaData().index(shard.index());
+                        if (canPerformOperationBasedRecovery(indexMetadata, primaryStore, currentNode, storeMetadata)) {
+                            return true;
```
> has a PRRL that matches the primary

Did you mean that we should only cancel the current recovery if the replacement is a noop recovery?

DaveCTurner replied:
> we should only cancel the current recovery if the replacement is a noop 

Yes.

(we can't be sure it will be a no-op recovery by the time it runs, but that's the best we can do here).

dnhatn replied:
Ok, thanks David. With this, we do not need to treat closed/frozen indices specially.

dnhatn committed:
Merge branch 'master' into replica-allocator

dnhatn committed:
feedback

dnhatn committed:
test name

dnhatn commented:
@henningandersen @DaveCTurner Thank you for reviewing. I have addressed your comments and suggestions in https://github.com/elastic/elasticsearch/pull/46959/commits/17bfb34157b69fa55b7c45086b5440796e67a7b4. Would you please take another look?

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java:
```diff
@@ -380,30 +372,87 @@ private static long computeMatchingBytes(TransportNodesListShardStoreMetaData.St
+        final long retainingSeqNoForPrimary = getRetainingSeqNoForNode(primaryStore, primaryNode);
+        final long retainingSeqNoForReplica = getRetainingSeqNoForNode(primaryStore, replicaNode);
+        final long matchingBytes = computeMatchingBytes(primaryStore, replicaStore);
+        final boolean isNoopRecovery = matchingBytes == Long.MAX_VALUE
```
can you change `computeMatchingBytes` not to do the sync id check anymore, but do that one here instead? That means that `matchingBytes` will truly mean the number of exact matching bytes.

ywelsch commented on server/src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java:
```diff
@@ -142,20 +146,20 @@ private StoreFilesMetaData listStoreMetaData(ShardId shardId) throws IOException
             //    reuse local resources.
             return new StoreFilesMetaData(shardId, Store.readMetadataSnapshot(shardPath.resolveIndex(), shardId,
-                nodeEnv::shardLock, logger));
+                nodeEnv::shardLock, logger), Collections.emptyList());
```
Should this not load the retention leases from disk? I know it's currently only relevant to have these if the shard is primary (and active), still a bit odd not to return those here without any comment.

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java:
```diff
@@ -52,29 +63,95 @@
 
-    public void testRecentPrimaryInformation() throws Exception {
+    /**
+     * Verify that if we found a new copy where it can perform a no-nop recovery,
```
no-nop 😀

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java:
```diff
@@ -52,29 +63,95 @@
                     .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1)
                     .put(IndexSettings.FILE_BASED_RECOVERY_THRESHOLD_SETTING.getKey(), 1.0f)
+                    .put(IndexService.RETENTION_LEASE_SYNC_INTERVAL_SETTING.getKey(), "100ms")
+                    .put(IndexService.GLOBAL_CHECKPOINT_SYNC_INTERVAL_SETTING.getKey(), "100ms")
```
This is only relevant when using async-durability, right?

dnhatn replied:
Yes.

ywelsch reviewed (commented):
I've taken a look to see how this works and left mainly smaller comments. Good stuff.

henningandersen commented on server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorTests.java:
```diff
@@ -369,6 +461,11 @@ public boolean getFetchDataCalledAndClean() {
         }
 
         public TestAllocator addData(DiscoveryNode node, String syncId, StoreFileMetaData... files) {
+            return addData(node, Collections.emptyList(), syncId, files);
```
Should we add random retention lease info here, to check that it is unused?

dnhatn replied:
We can break some existing tests if we add some random leases for the primary shard.

henningandersen commented on server/src/test/java/org/elasticsearch/index/store/StoreTests.java:
```diff
@@ -859,7 +860,7 @@ public void testStreamStoreFilesMetaData() throws Exception {
         TransportNodesListShardStoreMetaData.StoreFilesMetaData outStoreFileMetaData =
             new TransportNodesListShardStoreMetaData.StoreFilesMetaData(new ShardId("test", "_na_", 0),
-                metadataSnapshot);
+                metadataSnapshot, Collections.emptyList());
```
I think we should add something random here and validate that it is preserved when streaming.

henningandersen reviewed (commented):
Thanks @dnhatn, this is looking good. I left a number of smaller comments to address or comment on.

dnhatn committed:
Merge branch 'master' into replica-allocator

dnhatn committed:
yannick’s comments

dnhatn committed:
henning’s comments

dnhatn commented on server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java:
```diff
@@ -380,30 +372,87 @@ private static long computeMatchingBytes(TransportNodesListShardStoreMetaData.St
+        final long retainingSeqNoForReplica = getRetainingSeqNoForNode(primaryStore, replicaNode);
+        final long matchingBytes = computeMatchingBytes(primaryStore, replicaStore);
+        final boolean isNoopRecovery = matchingBytes == Long.MAX_VALUE
+            || (retainingSeqNoForReplica > 0 && retainingSeqNoForReplica == retainingSeqNoForPrimary);
```
+1 to use max_seq_no from the primary. I prefer to make this change in a follow-up as we might need to iterate on how we snapshot max_seq_no. I made the change "retainingSeqNoForReplica >= retainingSeqNoForPrimary".

dnhatn commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2619,6 +2619,13 @@ public void removePeerRecoveryRetentionLease(String nodeId, ActionListener<Repli
+    /**
+     * Returns a list of retention leases for peer recovery installed in this shard copy.
+     */
+    public List<RetentionLease> getPeerRecoveryRetentionLeases() {
```
I made this change, but I was not happy as we have to use "substring" to extract nodeId from the retention lease Id. Let me know if you are okay with substring. Anyway, I moved `getRetainingSeqNoForNode` to `StoreFilesMetaData`.

henningandersen replied:
OK, thanks for trying it out. I am OK with this as is, it was definitely a smaller thing.

dnhatn commented:
@ywelsch @henningandersen Thank you for another helpful review. I have responded/addressed your comments. Would you please take another look?

dnhatn commented:
Although [the failure](https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/8483/) is from a newly introduced test, I think this PR is still ready for another round. I am investigating the test failure.

dnhatn committed:
use persistent settings

dnhatn committed:
Merge branch 'master' into replica-allocator

dnhatn commented:
I pushed https://github.com/elastic/elasticsearch/pull/46959/commits/2ff795f4fe7bc7acddfd0461ef8da7d13d3344d6 to use persistent settings instead of transient settings before restarting the cluster. This commit should fix the above test failure.

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java:
```diff
@@ -355,21 +355,20 @@ private MatchingNodes findMatchingNodes(ShardRouting shard, RoutingAllocation al
+        return sizeMatched;
+    }
+
+    private static boolean hasMatchedSyncId(TransportNodesListShardStoreMetaData.StoreFilesMetaData primaryStore,
```
hasMatchingSyncId

ywelsch reviewed (approved):
Thanks Nhat!

dnhatn committed:
grammar :)

henningandersen reviewed (approved):
LGTM.

Thanks @dnhatn

DaveCTurner commented on server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java:
```diff
@@ -41,8 +50,10 @@
+import static org.hamcrest.Matchers.not;
 
 @ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
 public class ReplicaShardAllocatorIT extends ESIntegTestCase {
```
I think there would be value to an `ESIntegTestCase` showing that we pick a copy suitable for ops-based recovery in a full-cluster restart in which there was ongoing indexing, as well as these tests of the behaviour of no-op recoveries.

dnhatn replied:
Yes, I add a test in https://github.com/elastic/elasticsearch/pull/46959/commits/a299c0f16a6421d0ea1b2e8481f3a1864ef320ce.

DaveCTurner reviewed (commented):
Thanks @dnhatn and apologies for the delayed review. I left a few questions, but no blockers.

dnhatn committed:
Merge branch 'master' into replica-allocator

dnhatn committed:
retainingSeqNoForPrimary >= 0

dnhatn committed:
add test that prefers copy with highest matching operations

dnhatn committed:
do not check for empty store again

dnhatn commented on server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java:
```diff
@@ -332,52 +320,80 @@ private MatchingNodes findMatchingNodes(ShardRouting shard, RoutingAllocation al
+        }
+        final long retainingSeqNoForPrimary = primaryStore.getPeerRecoveryRetentionLeaseRetainingSeqNo(primaryNode);
+        final long retainingSeqNoForReplica = primaryStore.getPeerRecoveryRetentionLeaseRetainingSeqNo(replicaNode);
+        final boolean isNoopRecovery = (retainingSeqNoForReplica >= 0 && retainingSeqNoForReplica >= retainingSeqNoForPrimary)
```
Good point. I pushed https://github.com/elastic/elasticsearch/pull/46959/commits/c744a762775fdaf631980e8b5bc6491e6e023371.

dnhatn commented on server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java:
```diff
@@ -332,52 +320,80 @@ private MatchingNodes findMatchingNodes(ShardRouting shard, RoutingAllocation al
+        DiscoveryNode primaryNode, TransportNodesListShardStoreMetaData.StoreFilesMetaData primaryStore,
+        DiscoveryNode replicaNode, TransportNodesListShardStoreMetaData.StoreFilesMetaData replicaStore) {
+        if (replicaStore.isEmpty()) {
+            return new MatchingNode(0, -1, false); // store is corrupted
```
Fixed in https://github.com/elastic/elasticsearch/pull/46959/commits/dc7fb14da15992659d06052ec29a11fb7cdbaa20.

dnhatn commented:
@DaveCTurner Thanks for looking. I have addressed your comments. Would you please take another look?

DaveCTurner reviewed (approved):
LGTM thanks @dnhatn

dnhatn commented:
@henningandersen @DaveCTurner @ywelsch Thank you very much for your helpful reviews.
